src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,758 @@
+        header_i++;
+    }
+
+    size_t wrote = 0;
+    while (this->got_len != wrote)
+    {
+        const size_t space_on_page = XLOG_BLCKSZ - bufUsed(output) % XLOG_BLCKSZ;
+        size_t to_write = Min(space_on_page, this->got_len - wrote);
+        checkOutputSize(output, to_write);
+
+        memcpy(bufRemainsPtr(output), ((char *) this->record) + wrote, to_write);
+        wrote += to_write;
+
+        bufUsedInc(output, to_write);
+
+        // write header
+        if (header_i < lstSize(this->headers))
+        {
+            to_write = SizeOfXLogShortPHD;
+            checkOutputSize(output, to_write);
+            memcpy(bufRemainsPtr(output), lstGet(this->headers, header_i), to_write);
+            bufUsedInc(output, to_write);
+
+            header_i++;
+        }
+    }
+
+    const size_t align_size = MAXALIGN(this->got_len) - this->got_len;
+    checkOutputSize(output, align_size);
+    memset(bufRemainsPtr(output), 0, align_size);
+    bufUsedInc(output, align_size);
+
+    this->got_len = 0;
+}
+
+static bool
+readBeginOfRecord(WalFilterState *const main_state)
+{
+    bool result = false;
+    MEM_CONTEXT_TEMP_BEGIN();
+
+    const String *walSegment = NULL;
+    const TimeLineID timeLine = main_state->current_header->xlp_tli;
+    uint64_t segno = main_state->current_header->xlp_pageaddr / XLOG_SEG_SIZE;
+    const String *const path = strNewFmt("%08X%08X", timeLine, (uint32) (segno / XLogSegmentsPerXLogId));
+
+    const StringList *const segmentList = storageListP(
+        storageRepoIdx(main_state->archiveInfo->repoIdx),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(path)),
+        .expression = strNewFmt("^[0-f]{24}-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$"));
+
+    if (strLstEmpty(segmentList))
+    {
+        // an exotic case where we couldn't even find the current file.
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segno_diff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *const file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo);
+
+        if (segno - fileSegNo < segno_diff && fileSegNo < segno)
+        {
+            segno_diff = segno - fileSegNo;
+            walSegment = strLstGet(segmentList, i);
+        }
+    }
+
+    // current file is oldest
+    if (segno_diff == UINT64_MAX)
+    {
+        memContextSwitchBack();
+        memContextDiscard();
+        return false;
+    }
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(walSegment)));
+
+    // If there is a cipher then add the decrypt filter
+    if (main_state->archiveInfo->cipherType != cipherTypeNone)
+    {
+        ioFilterGroupAdd(
```

Comment:
could we avoid copy-paste `ioFilterGroupAdd` for descrypt and decompress and other similar logic from `archiveGetFile` to `readBeginOfRecord` and `getEndOfRecord`?